### PR TITLE
scipy.stats: rv_histogram fit method

### PR DESCRIPTION
**Pull request by yrobink**
#### Reference issue
No reference ?

#### What does this implement/fix?
All continuous distribution can be defined with syntax
"dist( *dist.fit(data) )", except rv_histogram. I propose
a fit method for rv_histogram. An example has been
added to the documentation of rv_histogram class.

#### Additional information
Already implemented in "SBCK.tools.rv_histogram" class,
see the file "python/SBCK/tools/__rv_extend.py" of the 
repository  "https://github.com/yrobink/SBCK".

### CODE REVIEW

File: scipy/stats/_continuous_distns.py [hunk 1]
```diff
@@ -8501,7 +8504,18 @@ def _updated_ctor_param(self):
         dct = super(rv_histogram, self)._updated_ctor_param()
         dct['histogram'] = self._histogram
         return dct
-
+    
+    def fit( data , **kwds ):
```
Comment (tylerjereddy) [inline scipy/stats/_continuous_distns.py:8508]: looks like a few other `fit()` methods implemented for subclasses of `rv_continuous` have a call signature like this:

`fit(self, data, *args, **kwds):`

It looks like the unit test machinery that iterates over distributions for testing is expecting a signature like that and causing a failure here.

**Review by tylerjereddy — COMMENTED** [review]
Can't really comment on the domain-specific `stats` stuff, but I added a few comments based on the CI failures and what I can see in the code in case it helps.

**Comment by mdhaber** [comment]
@yrobink Although `rv_histogram` technically inherits from `rv_continuous`, it is not supposed used like the other `rv_continuous` distributions, and so it appears alongside `rv_discrete` and `rv_continuous` at the top of the `stats` documentation:
image

With any other distribution, you can call the methods `pdf`, `cdf`, etc... like:
```
> from scipy.stats import norm as dist
> print(dist.pdf(1))
0.24197072451914337
```
For `rv_continuous`, you get an error:
```
> from scipy.stats import rv_continuous as dist
> print(dist.pdf(1))

Traceback (most recent call last):

  File "<ipython-input-44-4c0bb8ea87ba>", line 2, in <module>
    print(dist.pdf(1))

TypeError: pdf() missing 1 required positional argument: 'x'
```

The documentation shows how it's supposed to be used, e.g.
```
> import scipy.stats
> import numpy as np
> data = scipy.stats.norm.rvs(size=100000, loc=0, scale=1.5, random_state=123)
> hist = np.histogram(data, bins=100)
> dist = scipy.stats.rv_histogram(hist)  # hist_dist behaves like an rv_continuous
> print(dist.pdf(1))
0.20538577847618705
```
It _does_ have a `fit` method that returns the best fit loc and scale. 
```
> data = scipy.stats.norm.rvs(size=100000, loc=2, scale=10, random_state=123)
> dist.fit(data)
(2.3567669471146084, 7.446631756549467)
```
Do you have a suggestion for how the documentation could be more clear about that?

**Comment by mdhaber** [comment]
@yrobink do you agree that this is not how `rv_histogram` is intended to be used? Should we maybe improve the documentation rather than adopting this PR?

**Comment by yrobink** [comment]
@mdhaber When you write:

```
> from scipy.stats import rv_continuous as dist
> print(dist.pdf(1))
```

What is masked is that you are actually calling:

```
> from scipy.stats import rv_continuous as dist
> print(dist.pdf( 1 , loc = 0 , scale = 1 ))
```

All methods `pdf`, `cdf`... takes as default arguments `loc=0` and `scale=1` for `scipy.stats.norm`, but it is not true for all distributions, e.g. `scipy.stats.beta.pdf(1)` raises a `TypeError` (the parameters `a` and `b` must be given). Here `scipy.stats.norm` has these default arguments because (I think):

```
norm(loc=0,scale=1) = (norm( loc = loc , scale = scale ) - loc )/ scale
```

But because you do not have a similar result for `scipy.stats.beta`, you do not have default arguments. For the particular case of `scipy.stats.rv_histogram`, this kind of distribution is called "non parametric" because there are no parameters that describe the distribution, it is entirely defined by the quantiles of a data set (inferred by `np.histogram`), and, in a sense, the return  of `np.histogram` plays the role of parameters `loc` / `scale` or `a` / `b`.

The idea of my PR is that in the same way as each distribution return the parameters fitted from a dataset, a `fit` method of `rv_histogram` should return a `np.histogram`. Furthermore, the documentation is slightly ambiguous, because each distribution must be set with their own parameters (even if it is hidden for `scipy.stats.norm`).

My goal, and I underestimated the difficulty, is to propose a PR for `rv_histogram` where it is clear that `np.histogram` is the parameter, and can be fit with the fit method.  With this approach, all distributions of `scipy.stats` could be defined from a dataset with syntax:

```
> ## dist = any scipy.stats.*
> dist_infered_from_data = dist( *dist.fit(data) )
```

For the moment it's a work in progress on my side to understand how `scipy.stats` is built.

Don't hesitate if I haven't been clear,

**Comment by mdhaber** [comment]
> The idea of my PR is that in the same way as each distribution return the parameters fitted from a dataset, **a fit method of rv_histogram should return a np.histogram**. 

@yrobink I think I understand what you are trying to do; let me try to rephrase so you can confirm or deny:

The documentation of `rv_histogram` stats that
> The pdf is defined as a stepwise function from the provided histogram.

So you want the `fit` method to treat the height of each step as a shape parameter, and you want `fit` to find the values of the height of each step that best fits the data. Is that correct?

If so, the problem is that here is the full sentence from the `rv_histogram` documentation:
> **There are no additional shape parameters except for the loc and scale.** The pdf is defined as a stepwise function from the provided histogram

So the heights of the steps are _not_ parameters of the `rv_histogram`. If they are not parameters, then the `fit` method does not fit them. I think what you are asking for is reasonable - it's just not the way that the `rv_histogram` class is supposed to work. If it _were_ supposed to work that way, then you'd have to pass a `np.histogram` into, e.g., the `pdf` function every time you call it, just like you have to pass `a` and `b` into `beta.pdf`.

**Comment by WarrenWeckesser** [comment]
I agree with @mdhaber's points; this version of `fit` is not appropriate for `rv_histogram`.

The `rv_histogram` class is designed such that the given histogram is a fixed part of the distribution; the counts and bin edges are not parameters of the distribution in the same way that, say, `a`, `b`, `loc` and `scale` are parameters of the `beta` distribution.  The only parameters available for fitting an instance of `rv_histogram` are `loc` and `scale` (and I doubt doing so would be useful).   The `fit` method is specifically documented as being the *maximum likelihood estimate* of the distribution parameters to the given data.  Generating a new histogram from the given data is just not what `fit` is designed to do.

@yrobink, thanks for working on this, but this PR is not implementing the correct behavior for the `fit` method, so I'm closing the PR.

**Pull request closed without merging**